libraries/freertos_plus/standard/freertos_plus_tcp/source/FreeRTOS_TCP_IP.c

```diff
@@ -3031,216 +3073,219 @@ const IPHeader_t *pxIPHeader;
+					if( pxSocket->u.xTCP.ucTCPState == ( uint8_t ) eCONNECT_SYN )
 					{
-						/* Send a challenge ACK. */
-						( void ) prvTCPSendChallengeAck( pxNetworkBuffer );
+						/* Per the above RFC, "In the SYN-SENT state ... the RST is
+						acceptable if the ACK field acknowledges the SYN." */
+						if( ulAckNumber == ( pxSocket->u.xTCP.xTCPWindow.ulOurSequenceNumber + 1UL ) )
+						{
+							vTCPStateChange( pxSocket, eCLOSED );
+						}
 					}
 					else
 					{
-						/* Nothing. */
+						/* Check whether the packet matches the next expected sequence number. */
+						if( ulSequenceNumber == pxSocket->u.xTCP.xTCPWindow.rx.ulCurrentSequenceNumber )
+						{
+							vTCPStateChange( pxSocket, eCLOSED );
+						}
+						/* Otherwise, check whether the packet is within the receive window. */
+						else if( ( ulSequenceNumber > pxSocket->u.xTCP.xTCPWindow.rx.ulCurrentSequenceNumber ) &&
+								 ( ulSequenceNumber < ( pxSocket->u.xTCP.xTCPWindow.rx.ulCurrentSequenceNumber +
+													  pxSocket->u.xTCP.xTCPWindow.xSize.ulRxWindowLength ) ) )
+						{
+							/* Send a challenge ACK. */
+							( void ) prvTCPSendChallengeAck( pxNetworkBuffer );
+						}
+						else
+						{
+							/* Nothing. */
+						}
 					}
-				}
 
-				/* Otherwise, do nothing. In any case, the packet cannot be handled. */
-				xResult = pdFAIL;
-			}
-			else if( ( ( ucTCPFlags & tcpTCP_FLAG_CTRL ) == tcpTCP_FLAG_SYN ) && ( pxSocket->u.xTCP.ucTCPState >= ( uint8_t ) eESTABLISHED ) )
-			{
-				/* SYN flag while this socket is already connected. */
-				FreeRTOS_debug_printf( ( "TCP: SYN unexpected from %lxip:%u\n", ulRemoteIP, xRemotePort ) );
+					/* Otherwise, do nothing. In any case, the packet cannot be handled. */
+					xResult = pdFAIL;
+				}
+				else if( ( ( ucTCPFlags & tcpTCP_FLAG_CTRL ) == tcpTCP_FLAG_SYN ) && ( pxSocket->u.xTCP.ucTCPState >= ( uint8_t ) eESTABLISHED ) )
+				{
+					/* SYN flag while this socket is already connected. */
+					FreeRTOS_debug_printf( ( "TCP: SYN unexpected from %lxip:%u\n", ulRemoteIP, xRemotePort ) );
 
-				/* The packet cannot be handled. */
-				xResult = pdFAIL;
-			}
-			else
-			{
-				/* Update the copy of the TCP header only (skipping eth and IP
-				headers).  It might be used later on, whenever data must be sent
-				to the peer. */
-				const BaseType_t lOffset = ipNUMERIC_CAST( BaseType_t, ipSIZE_OF_ETH_HEADER + uxIPHeaderSizeSocket( pxSocket ) );
-				( void ) memcpy( &( pxSocket->u.xTCP.xPacket.u.ucLastPacket[ lOffset ] ),
-								 &( pxNetworkBuffer->pucEthernetBuffer[ lOffset ] ),
-								 ipSIZE_OF_TCP_HEADER );
+					/* The packet cannot be handled. */
+					xResult = pdFAIL;
+				}
+				else
+				{
+					/* Update the copy of the TCP header only (skipping eth and IP
+					headers).  It might be used later on, whenever data must be sent
+					to the peer. */
+					const UBaseType_t lOffset = ipNUMERIC_CAST( UBaseType_t, ipSIZE_OF_ETH_HEADER + uxIPHeaderSizeSocket( pxSocket ) );
```

Comment:
lOffset is purely used as an array index, shall we change it to `const size_t uxOffset`?